Version Bump: Add CLI utility

psealock opened the pull request:
### All Submissions:

-   [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
-   [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
-   [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

Contributes to https://github.com/woocommerce/woocommerce/issues/34424

### How to test the changes in this Pull Request:

1. Try bumping to an invalid version `ts-node ./tools/version-bump/index.ts bump woocommerce -v 99999` and see an appropriate error occur.
2. Try bumping to a lower version than current `ts-node ./tools/version-bump/index.ts bump woocommerce -v 6.8.0-beta.1` and see an appropriate error occur.
3. Bump a dev version as if you were updating trunk after a code freeze, `ts-node ./tools/version-bump/index.ts bump woocommerce -v 7.1.0-dev`. Check the diff and note that all updated files will get a regular version `7.1.0` while the main plugin file gets `7.1.0-dev`. The readme top level changelog title also gets updated.
4. Checkout all files to start fresh and bump a minor version `ts-node ./tools/version-bump/index.ts bump woocommerce -v 7.1.1`. Check the diff to make sure changes are appropriate.
5. Checkout all files again and bump to a prerelease version `ts-node ./tools/version-bump/index.ts bump woocommerce -v 7.1.0-beta.2`. 

**Questions**
1. Do the update make sense in each case? 
2. Do the testing instructions correctly assume desired behaviour?
3. So long as the user increases a version, anything goes. Is checking for appropriate bumps worthwhile? I don't think its worth the trouble, especially if we automate this with a Github workflow.

### Other information:

-   [x] Have you added an explanation of what your changes do and why you'd like us to include them?
-   [x] Have you written new tests for your changes, as applicable?
-   [x] Have you successfully run tests with your changes locally?
-   [x] Have you created a changelog file for each project being changed, ie `pnpm changelog add --filter=<project>`?

<!-- Mark completed items with an [x] -->

### FOR PR REVIEWER ONLY:

-   [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

psealock committed:
Add version-bump command

psealock committed:
dev version

psealock committed:
check prereleaseParameters

psealock committed:
class file

psealock committed:
better regex

psealock committed:
correct write file

psealock committed:
update readme.txt

psealock committed:
validateFn

psealock committed:
pass plugin arg

psealock committed:
update getCurrentVersion

psealock committed:
add docblocks

psealock committed:
bail if no class file

samueljseay commented:
@psealock mentioning this here for visibility, but we are supposed to be building new CLI tools in Commander.js not oclif. I've got some good examples in my recent work here:

https://github.com/woocommerce/woocommerce/tree/dev/analyzer-lint

and

in this PR here: https://github.com/woocommerce/woocommerce/pull/34496

psealock committed:
switch to commander.js

psealock committed:
update functions

psealock committed:
cleanup

psealock committed:
remove unused files

psealock committed:
changelog

psealock commented:
Thanks for the note @samueljseay, I have just moved this over to use `commanderjs` and updated the instructions.

In https://github.com/woocommerce/woocommerce/pull/34496 you mention making the Logger functionality useable in other contexts, so here I added a very basic Logger class as a placeholder until that happens.

jonathansadowski commented:
Thanks for grabbing this @psealock... this'll save us some headaches in future releases for sure.

A few questions / comments:

If we're going to start writing CLI tools in TS.., should we add ts-node as a dev dependency so that we can `pnpx ts-node...`?

> Bump a dev version as if you were updating trunk after a code freeze, ts-node ./tools/version-bump/index.ts bump woocommerce -v 7.1.0-dev. Check the diff and note that all updated files will get a regular version 7.1.0 while the main plugin file gets 7.1.0-dev. The readme top level changelog title also gets updated.

I don't think we should update the stable tag in this case. 7.1.0 isn't the correct stable, as when the release branch eventually gets cut from `trunk`, the most-likely stable version at the time will be 7.0.0 (or, perhaps if there was a fix release, 7.0.1, for example). I honestly don't know the best way to handle the stable tag re: this tool... might be worthy of discussion

> Checkout all files again and bump to a prerelease version ts-node ./tools/version-bump/index.ts bump woocommerce -v 7.1.0-beta.2. The only file to change will be the main file

I do wonder if in this case the behavior should be the same as bumping to `7.1.0-dev`... that-is, if any of the other files need updating, they get updated... but if they were previously prepared correctly, the next result is that only the single file gets updated.

samueljseay commented on tools/version-bump/lib/update.ts:
```diff
@@ -0,0 +1,136 @@
+): void => {
+	const filePath = `plugins/${ plugin }/${ plugin }.php`;
+	try {
+		const pluginFileContents = readFileSync( filePath, 'utf8' );
```
Can we move away from any sync blocking tasks? It makes CLI tasks run much slower and also makes things like progress indicators not work.

If you change this to async and use `await` by simply importing readFile like:

```
import { promises } from 'fs';


...later

await promises.readFile(...)
`

samueljseay commented on tools/version-bump/lib/update.ts:
```diff
@@ -0,0 +1,136 @@
+			/Version: \d.\d.\d.*\n/m,
+			`Version: ${ nextVersion }\n`
+		);
+		writeFileSync( filePath, updatedPluginFileContents );
```
same here, if we use writeFile from `promises` we don't have to block.

samueljseay commented on tools/version-bump/lib/validate.ts:
```diff
@@ -0,0 +1,56 @@
+export const getCurrentVersion = ( plugin: string ): string | void => {
+	try {
+		const composerJSON = JSON.parse(
+			readFileSync( `plugins/${ plugin }/composer.json`, 'utf8' )
```
prefer async here

samueljseay commented on tools/version-bump/lib/update.ts:
```diff
@@ -0,0 +1,136 @@
+): void => {
+	const filePath = `plugins/${ plugin }/includes/class-${ plugin }.php`;
+
+	if ( ! existsSync( filePath ) ) {
```
prefer async here

samueljseay commented on tools/version-bump/lib/update.ts:
```diff
@@ -0,0 +1,136 @@
+	}
+
+	try {
+		const classPluginFileContents = readFileSync( filePath, 'utf8' );
```
prefer async here

samueljseay commented on tools/version-bump/lib/update.ts:
```diff
@@ -0,0 +1,136 @@
+			`public $version = '${ nextVersion }';\n`
+		);
+
+		writeFileSync( filePath, updatedClassPluginFileContents );
```
prefer async here.

samueljseay commented on tools/version-bump/lib/update.ts:
```diff
@@ -0,0 +1,136 @@
+): void => {
+	const filePath = `plugins/${ plugin }/composer.json`;
+	try {
+		const composerJson = JSON.parse( readFileSync( filePath, 'utf8' ) );
```
prefer async here.

samueljseay commented on tools/version-bump/lib/update.ts:
```diff
@@ -0,0 +1,136 @@
+	try {
+		const composerJson = JSON.parse( readFileSync( filePath, 'utf8' ) );
+		composerJson.version = nextVersion;
+		writeFileSync(
```
prefer async here.

samueljseay commented on tools/version-bump/lib/update.ts:
```diff
@@ -0,0 +1,136 @@
+			`= ${ nextVersion } ${ new Date().getFullYear() }-XX-XX =\n`
+		);
+
+		writeFileSync( filePath, updatedReadmeContents );
```
prefer async here.

samueljseay commented on tools/version-bump/lib/update.ts:
```diff
@@ -0,0 +1,136 @@
+): void => {
+	const filePath = `plugins/${ plugin }/readme.txt`;
+	try {
+		const readmeContents = readFileSync( filePath, 'utf8' );
```
prefer async here.

samueljseay commented on tools/version-bump/lib/update.ts:
```diff
@@ -0,0 +1,136 @@
+): void => {
+	const filePath = `plugins/${ plugin }/readme.txt`;
+	try {
+		const readmeContents = readFileSync( filePath, 'utf8' );
```
prefer async here

samueljseay commented on tools/version-bump/lib/validate.ts:
```diff
@@ -0,0 +1,56 @@
+
+	if ( ! currentVersion ) {
+		Logger.error( 'Unable to determine current version' );
+	} else if ( versionLessThan( nextVersion, currentVersion ) ) {
```
Love seeing you're using semver too. makes all of this version checking so much nicer than what we've had in the past.

samueljseay commented on tools/version-bump/tsconfig.json:
```diff
@@ -0,0 +1,17 @@
+{
+	"compilerOptions": {
```
My thought and what I've been doing in other cli tools is to disable any compilation for now. I was thinking we could work towards compilation as an enhancement later if we want to bake any of our code as standalone.

you can also use the `@tsconfig` module to get nice defaults for node 16 

e.g. this is my tsconfig in release post generator:

```
{
  "extends": "@tsconfig/node16/tsconfig.json",
  "ts-node": {
    "transpileOnly": true,
    "files": true,
  }
}
```

The only issue I had is with some pure ESM modules because there isn't really good support in node/TS for mixing pure ESM with common JS. Occasionally I had to drop down to an older non esm module in some cases.

samueljseay reviewed (commented):
This is looking good but I would like to see us remove all blocking/sync calls from our tools. Node has most of the utils baked in these days that you can async everything. In this case its all calls to file APIs which already has a promises alternative but for others you can use promisify in future to avoid blocking: https://nodejs.org/api/util.html#utilpromisifyoriginal

psealock committed:
update stable tag only on normal bumps

psealock committed:
redo logic

psealock committed:
whitespace

psealock commented:
Thanks for the review @jonathansadowski,

I changed the logic such that updating with `-dev` does not change the stable tag and updating with a prerelease (ie, `-beta.1`) updates all the files, which as you mentioned, won't be a change if the previous version bump was done correctly.

psealock committed:
updated tsconfig

psealock committed:
updatePluginFile async

psealock committed:
async all the things

psealock committed:
Update fs exists to async

psealock commented:
@samueljseay I have updated all synchronous calls to async.

Are you planning on making the Logger class accessible to all the tools? If so, I'll wait and adapt this PR when then becomes available to enable better logging and spinners.

psealock committed:
Update package.json as well

psealock commented:
Updating `package.json` was missing, so I added that in https://github.com/woocommerce/woocommerce/pull/34555/commits/427629081c383347cc6b24d6ffc7a1679a305f16

samueljseay commented:
> Are you planning on making the Logger class accessible to all the tools? If so, I'll wait and adapt this PR when then becomes available to enable better logging and spinners.

Yes, thats coming in the branch I'm working on now. It will be pretty simple to drop in.

samueljseay commented on tools/version-bump/index.ts:
```diff
@@ -0,0 +1,8 @@
+import './commands/bump';
+
+// Start the program
+program.parse();
```
Pretty sure you need to pass `process.argv` here.

psealock replied:
I noticed `release-post-generator` didn't pass them in either, but [that doesn't seem to matter](https://github.com/tj/commander.js#common-option-types-boolean-and-value)

>program.parse(arguments) processes the arguments, leaving any args not consumed by the program options in the program.args array. The parameter is optional and defaults to process.argv.

I just passed them in anyways for better clarity.

samueljseay replied:
Oh you're right! I missed that it was optional!

samueljseay commented on tools/version-bump/lib/update.ts:
```diff
@@ -0,0 +1,141 @@
+	try {
+		await stat( filePath );
+	} catch ( e ) {
+		// Class file does not exist, return early.
```
If the class file doesn't exist is this an expected error? should this throw?

psealock replied:
No it should not throw an error. Only core has the `class-woocommerce.php` file with a version requiring update, so I want to return early if it does not exist.

I wanted to do this, but `exist` is [deprecated](https://nodejs.org/api/fs.html#fsexistspath-callback) so I need to use `stat` or `access` both of which return errors when the file does not exist.

```js
if ( exists( filePath ) { 
    return;
}
```

samueljseay commented on tools/version-bump/lib/update.ts:
```diff
@@ -0,0 +1,141 @@
+export const updateClassPluginFile = async (
+	plugin: string,
+	nextVersion: string
+): Promise< void > => {
```
Just a heads up, TypeScript is very good at inferring the return type so unless you want to make it explicit what's being returned you don't need to add the return type.

psealock replied:
Ok, good to know. Since its in there, I'll leave it because it makes it apparent.

samueljseay commented on tools/version-bump/lib/update.ts:
```diff
@@ -0,0 +1,141 @@
+	type: 'package' | 'composer',
+	plugin: string,
+	nextVersion: string
+): Promise< void > => {
```
Again return type is not needed.

samueljseay commented on tools/version-bump/lib/update.ts:
```diff
@@ -0,0 +1,141 @@
+): Promise< void > => {
+	const filePath = `plugins/${ plugin }/${ type }.json`;
+	try {
+		const composerJson = JSON.parse( await readFile( filePath, 'utf8' ) );
```
Is this a safe path? shouldn't this be absolute path from the top of `woocommerce` or something? This would only work if you run from the correct directory otherwise

psealock replied:
I updated all the paths to be more robust in https://github.com/woocommerce/woocommerce/pull/34555/commits/8c4358e4d37426203c36e9582aef910448d20dbc

samueljseay commented on tools/version-bump/lib/update.ts:
```diff
@@ -0,0 +1,141 @@
+): Promise< void > => {
+	const filePath = `plugins/${ plugin }/${ plugin }.php`;
+	try {
+		const pluginFileContents = await readFile( filePath, 'utf8' );
```
Again just checking, these relative paths seem a bit fragile

samueljseay commented on tools/version-bump/lib/validate.ts:
```diff
@@ -0,0 +1,82 @@
+): Promise< string | void > => {
+	try {
+		const composerJSON = JSON.parse(
+			await readFile( `plugins/${ plugin }/composer.json`, 'utf8' )
```
again relative path

samueljseay commented on tools/version-bump/lib/validate.ts:
```diff
@@ -0,0 +1,82 @@
+		return `${ str }.${ value }`;
+	}, '' );
+
+	return prereleaseVersion.replace( substr.toString(), '' );
```
I wonder if this version parsing could be done with `semver`. Its fine how it is, just food for thought.

psealock replied:
I scoured the docs for a method that does this, ie takes `7.1.0-beta.1` and returns `7.1.0` but I could not find anything! In the end, I opted to change this to use `parse` which breaks it down into an object and concatenate major, minor, and patch. Less logic seems better.

samueljseay reviewed (commented):
Just a few comments, questions and possibly nitpicks from me. Let me know what you think

psealock committed:
use semvers parse to remove prerelease params

psealock committed:
less flakiy file paths

psealock committed:
pass in process.argv

samueljseay reviewed (dismissed):
Changes look good to me!

psealock committed:
pnpm lock file update
